# lab2

In [4]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv ('s3://ia340-fall2025-clark4jw-sldfkjsalgjsdoh/Crime_Incidents_in_2025.csv') #loading my data from S3
df.head(10) #show top 10 rows

X            Y       CCN              REPORT_DAT     SHIFT  \
0  400588.4650  129470.0424   5131706  2025/02/05 05:00:00+00  MIDNIGHT   
1  405684.5761  135633.8979  10133557  2025/06/30 04:00:00+00  MIDNIGHT   
2  397581.8300  141846.4300  25079053  2025/05/27 22:19:32+00   EVENING   
3  399215.8700  138751.2000  25079273  2025/08/11 17:39:31+00       DAY   
4  397700.5000  137976.2500  25421607  2025/05/05 18:31:00+00       DAY   
5  396525.4500  138714.0600  25421619  2025/05/05 20:04:15+00   EVENING   
6  400495.3800  130841.2700  25079455  2025/05/28 17:17:45+00       DAY   
7  397171.1100  137408.2500  25079796  2025/05/29 07:24:52+00  MIDNIGHT   
8  396917.7000  138058.7800  25421322  2025/04/15 23:32:40+00   EVENING   
9  395366.7500  140009.3600  25421413  2025/04/23 12:31:16+00       DAY   

   METHOD      OFFENSE                                      BLOCK  \
0     GUN     HOMICIDE         800 - 899 BLOCK OF XENIA STREET SE   
1     GUN     HOMICIDE           4600 - 4699 BLOCK OF A STREET SE   
2  OTHERS      ROBBERY        4400 - 4499 BLOCK OF IOWA AVENUE NW   
3  OTHERS      ROBBERY  1800 - 2199 BLOCK OF NORTH CAPITOL STREET   
4  OTHERS  THEFT/OTHER           1000 - 1099 BLOCK OF P STREET NW   
5  OTHERS  THEFT/OTHER     1700 - 1799 BLOCK OF WILLARD STREET NW   
6  OTHERS  THEFT/OTHER        800 - 899 BLOCK OF HEIGHTS ALLEY SE   
7  OTHERS  THEFT/OTHER     1100 - 1199 BLOCK OF VERMONT AVENUE NW   
8  OTHERS  THEFT/OTHER      1500 - 1599 BLOCK OF CHURCH STREET NW   
9  OTHERS  THEFT/OTHER   2300 - 2399 BLOCK OF CATHEDRAL AVENUE NW   

          XBLOCK         YBLOCK  ...  BLOCK_GROUP CENSUS_TRACT  \
0  400588.464990  129470.042358  ...     009801 1       9801.0   
1  405684.576067  135633.897851  ...     009906 1       9906.0   
2  397581.830000  141846.430000  ...     002501 1       2501.0   
3  399215.870000  138751.200000  ...     003302 1       3302.0   
4  397700.500000  137976.250000  ...     004901 3       4901.0   
5  396525.450000  138714.060000  ...     004201 3       4201.0   
6  400495.380000  130841.270000  ...     010400 3      10400.0   
7  397171.110000  137408.250000  ...     010100 3      10100.0   
8  396917.700000  138058.780000  ...     005202 2       5202.0   
9  395366.750000  140009.360000  ...     000501 1        501.0   

   VOTING_PRECINCT   LATITUDE  LONGITUDE       BID              START_DATE  \
0     Precinct 121  38.833020 -76.993222       NaN  2005/09/27 07:25:00+00   
1     Precinct 104  38.888528 -76.934478       NaN  2010/09/15 00:45:00+00   
2      Precinct 48  38.944508 -77.027895       NaN  2025/05/27 21:11:00+00   
3      Precinct 19  38.916628 -77.009042       NaN  2025/05/28 07:40:00+00   
4      Precinct 18  38.909644 -77.026513       NaN  2025/05/04 14:45:00+00   
5     Precinct 141  38.916287 -77.040064       NaN  2025/04/23 18:33:00+00   
6     Precinct 119  38.845373 -76.994294       NaN  2025/05/28 16:34:00+00   
7      Precinct 17  38.904526 -77.032614  DOWNTOWN  2025/05/29 05:27:00+00   
8      Precinct 16  38.910386 -77.035539       NaN  2025/04/14 05:05:00+00   
9     Precinct 136  38.927950 -77.053434       NaN  2025/04/08 20:30:00+00   

                 END_DATE   OBJECTID  OCTO_RECORD_ID  
0  2005/09/27 07:26:00+00  786940029             NaN  
1  2010/09/14 04:00:00+00  786940034             NaN  
2  2025/05/27 22:18:00+00  786942870             NaN  
3  2025/05/28 07:45:00+00  786944170             NaN  
4  2025/05/04 14:50:00+00  786944171             NaN  
5  2025/04/23 18:34:00+00  786944172             NaN  
6  2025/05/28 17:17:00+00  786944274             NaN  
7  2025/05/29 06:00:00+00  786944275             NaN  
8  2025/04/14 05:05:00+00  786944276             NaN  
9  2025/04/22 14:15:00+00  786944277             NaN  

[10 rows x 25 columns]

In [5]:
df.shape #show number of rows and columns

(16612, 25)

In [6]:
df.info() #displays all columns and rows

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16612 entries, 0 to 16611
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   X                     16612 non-null  float64
 1   Y                     16612 non-null  float64
 2   CCN                   16612 non-null  int64  
 3   REPORT_DAT            16612 non-null  object 
 4   SHIFT                 16612 non-null  object 
 5   METHOD                16612 non-null  object 
 6   OFFENSE               16612 non-null  object 
 7   BLOCK                 16612 non-null  object 
 8   XBLOCK                16612 non-null  float64
 9   YBLOCK                16612 non-null  float64
 10  WARD                  16607 non-null  float64
 11  ANC                   16607 non-null  object 
 12  DISTRICT              16561 non-null  float64
 13  PSA                   16561 non-null  float64
 14  NEIGHBORHOOD_CLUSTER  16607 non-null  object 
 15  BLOCK_GROUP        

In [7]:
df.describe() #provides descriptive statistics of each column/variable

X              Y           CCN         XBLOCK  \
count   16612.000000   16612.000000  1.661200e+04   16612.000000   
mean   399223.899504  137978.638332  2.509500e+07  399223.899504   
std      3011.298002    3482.734790  5.500669e+05    3011.298002   
min    390519.560000  127300.240000  5.131706e+06  390519.560000   
25%    397229.100000  136474.240300  2.503431e+07  397229.100000   
50%    398916.585000  137848.510000  2.507004e+07  398916.585006   
75%    401014.009600  139965.547500  2.510658e+07  401014.009567   
max    407806.750900  147250.570000  8.927285e+07  407806.750917   

              YBLOCK          WARD      DISTRICT           PSA   CENSUS_TRACT  \
count   16612.000000  16607.000000  16561.000000  16561.000000   16601.000000   
mean   137978.638332      4.397844      3.681722    372.759133    7635.496295   
std      3482.734790      2.269673      1.825169    182.120965   35568.341322   
min    127300.240000      1.000000      1.000000    101.000000     101.000000   
25%    136474.240252      2.000000      2.000000    207.000000    3500.000000   
50%    137848.510000      5.000000      4.000000    401.000000    7304.000000   
75%    139965.547500      6.000000      5.000000    505.000000    9204.000000   
max    147250.570000      8.000000      7.000000    708.000000  980000.000000   

           LATITUDE     LONGITUDE      OBJECTID  OCTO_RECORD_ID  
count  16612.000000  16612.000000  1.661200e+04             0.0  
mean      38.909663    -77.008955  7.873447e+08             NaN  
std        0.031373      0.034721  1.073366e+05             NaN  
min       38.813473    -77.109338  7.869400e+08             NaN  
25%       38.896116    -77.031952  7.872527e+08             NaN  
50%       38.908480    -77.012499  7.872810e+08             NaN  
75%       38.927567    -76.988310  7.874623e+08             NaN  
max       38.993186    -76.910012  7.875041e+08             NaN

I have 16612 rows and 25 columns. My columns are floats, integers, and object columns. 

In [8]:
df = df.rename(columns={"NEIGHBORHOOD_CLUSTER": "Neighborhood"})  #renames column to easier to type name
df.head() #displays first 5 rows

X            Y       CCN              REPORT_DAT     SHIFT  \
0  400588.4650  129470.0424   5131706  2025/02/05 05:00:00+00  MIDNIGHT   
1  405684.5761  135633.8979  10133557  2025/06/30 04:00:00+00  MIDNIGHT   
2  397581.8300  141846.4300  25079053  2025/05/27 22:19:32+00   EVENING   
3  399215.8700  138751.2000  25079273  2025/08/11 17:39:31+00       DAY   
4  397700.5000  137976.2500  25421607  2025/05/05 18:31:00+00       DAY   

   METHOD      OFFENSE                                      BLOCK  \
0     GUN     HOMICIDE         800 - 899 BLOCK OF XENIA STREET SE   
1     GUN     HOMICIDE           4600 - 4699 BLOCK OF A STREET SE   
2  OTHERS      ROBBERY        4400 - 4499 BLOCK OF IOWA AVENUE NW   
3  OTHERS      ROBBERY  1800 - 2199 BLOCK OF NORTH CAPITOL STREET   
4  OTHERS  THEFT/OTHER           1000 - 1099 BLOCK OF P STREET NW   

          XBLOCK         YBLOCK  ...  BLOCK_GROUP CENSUS_TRACT  \
0  400588.464990  129470.042358  ...     009801 1       9801.0   
1  405684.576067  135633.897851  ...     009906 1       9906.0   
2  397581.830000  141846.430000  ...     002501 1       2501.0   
3  399215.870000  138751.200000  ...     003302 1       3302.0   
4  397700.500000  137976.250000  ...     004901 3       4901.0   

   VOTING_PRECINCT   LATITUDE  LONGITUDE  BID              START_DATE  \
0     Precinct 121  38.833020 -76.993222  NaN  2005/09/27 07:25:00+00   
1     Precinct 104  38.888528 -76.934478  NaN  2010/09/15 00:45:00+00   
2      Precinct 48  38.944508 -77.027895  NaN  2025/05/27 21:11:00+00   
3      Precinct 19  38.916628 -77.009042  NaN  2025/05/28 07:40:00+00   
4      Precinct 18  38.909644 -77.026513  NaN  2025/05/04 14:45:00+00   

                 END_DATE   OBJECTID  OCTO_RECORD_ID  
0  2005/09/27 07:26:00+00  786940029             NaN  
1  2010/09/14 04:00:00+00  786940034             NaN  
2  2025/05/27 22:18:00+00  786942870             NaN  
3  2025/05/28 07:45:00+00  786944170             NaN  
4  2025/05/04 14:50:00+00  786944171             NaN  

[5 rows x 25 columns]

In [9]:
df.isnull().sum() #examines which columns are missing data

X                      0
Y                      0
CCN                    0
REPORT_DAT             0
SHIFT                  0
METHOD                 0
OFFENSE                0
BLOCK                  0
XBLOCK                 0
YBLOCK                 0
WARD                   5
ANC                    5
DISTRICT              51
PSA                   51
Neighborhood           5
BLOCK_GROUP           11
CENSUS_TRACT          11
VOTING_PRECINCT        5
LATITUDE               0
LONGITUDE              0
BID                13327
START_DATE             2
END_DATE            1174
OBJECTID               0
OCTO_RECORD_ID     16612
dtype: int64

In [10]:
df = df[['X','Y','CCN','REPORT_DAT','SHIFT','METHOD','OFFENSE','BLOCK','XBLOCK','YBLOCK','LATITUDE','LONGITUDE']] # keep specific columns 
df.head()

X            Y       CCN              REPORT_DAT     SHIFT  \
0  400588.4650  129470.0424   5131706  2025/02/05 05:00:00+00  MIDNIGHT   
1  405684.5761  135633.8979  10133557  2025/06/30 04:00:00+00  MIDNIGHT   
2  397581.8300  141846.4300  25079053  2025/05/27 22:19:32+00   EVENING   
3  399215.8700  138751.2000  25079273  2025/08/11 17:39:31+00       DAY   
4  397700.5000  137976.2500  25421607  2025/05/05 18:31:00+00       DAY   

   METHOD      OFFENSE                                      BLOCK  \
0     GUN     HOMICIDE         800 - 899 BLOCK OF XENIA STREET SE   
1     GUN     HOMICIDE           4600 - 4699 BLOCK OF A STREET SE   
2  OTHERS      ROBBERY        4400 - 4499 BLOCK OF IOWA AVENUE NW   
3  OTHERS      ROBBERY  1800 - 2199 BLOCK OF NORTH CAPITOL STREET   
4  OTHERS  THEFT/OTHER           1000 - 1099 BLOCK OF P STREET NW   

          XBLOCK         YBLOCK   LATITUDE  LONGITUDE  
0  400588.464990  129470.042358  38.833020 -76.993222  
1  405684.576067  135633.897851  38.888528 -76.934478  
2  397581.830000  141846.430000  38.944508 -77.027895  
3  399215.870000  138751.200000  38.916628 -77.009042  
4  397700.500000  137976.250000  38.909644 -77.026513

I renamed "NEIGHBORHOOD_BLOCK" to "Neighborhood" for easy typing. I found many missing values, so I chose to keep the necessary columns that did not have missing values. 

In [11]:
df['CCN']

0         5131706
1        10133557
2        25079053
3        25079273
4        25421607
           ...   
16607    25420387
16608    25420731
16609    25421539
16610    25421842
16611    25421919
Name: CCN, Length: 16612, dtype: int64

In [20]:
df.sort_values(by='CCN') #sort values by CCN
df.head() #show first 5 rows

X            Y       CCN              REPORT_DAT     SHIFT  \
0  400588.4650  129470.0424   5131706  2025/02/05 05:00:00+00  MIDNIGHT   
1  405684.5761  135633.8979  10133557  2025/06/30 04:00:00+00  MIDNIGHT   
2  397581.8300  141846.4300  25079053  2025/05/27 22:19:32+00   EVENING   
3  399215.8700  138751.2000  25079273  2025/08/11 17:39:31+00       DAY   
4  397700.5000  137976.2500  25421607  2025/05/05 18:31:00+00       DAY   

   METHOD      OFFENSE                                      BLOCK  \
0     GUN     HOMICIDE         800 - 899 BLOCK OF XENIA STREET SE   
1     GUN     HOMICIDE           4600 - 4699 BLOCK OF A STREET SE   
2  OTHERS      ROBBERY        4400 - 4499 BLOCK OF IOWA AVENUE NW   
3  OTHERS      ROBBERY  1800 - 2199 BLOCK OF NORTH CAPITOL STREET   
4  OTHERS  THEFT/OTHER           1000 - 1099 BLOCK OF P STREET NW   

          XBLOCK         YBLOCK   LATITUDE  LONGITUDE  
0  400588.464990  129470.042358  38.833020 -76.993222  
1  405684.576067  135633.897851  38.888528 -76.934478  
2  397581.830000  141846.430000  38.944508 -77.027895  
3  399215.870000  138751.200000  38.916628 -77.009042  
4  397700.500000  137976.250000  38.909644 -77.026513

In [21]:
df.tail() #show last 5 rows

X          Y       CCN              REPORT_DAT    SHIFT  \
16607  406134.57  134786.65  25420387  2025/01/30 02:02:12+00  EVENING   
16608  399759.38  136609.78  25420731  2025/02/19 12:31:49+00      DAY   
16609  401297.97  136784.11  25421539  2025/04/29 23:01:35+00  EVENING   
16610  396307.96  137321.65  25421842  2025/05/15 21:02:18+00  EVENING   
16611  396330.15  139826.42  25421919  2025/05/21 21:03:07+00  EVENING   

       METHOD      OFFENSE                                 BLOCK     XBLOCK  \
16607  OTHERS  THEFT/OTHER     520 - 699 BLOCK OF 51ST STREET SE  406134.57   
16608  OTHERS  THEFT/OTHER        200 - 299 BLOCK OF F STREET NE  399759.38   
16609  OTHERS  THEFT/OTHER      1400 - 1425 BLOCK OF G STREET NE  401297.97   
16610  OTHERS  THEFT/OTHER      1800 - 1899 BLOCK OF L STREET NW  396307.96   
16611  OTHERS  THEFT/OTHER  1800 - 1899 BLOCK OF SUMMIT PLACE NW  396330.15   

          YBLOCK   LATITUDE  LONGITUDE  
16607  134786.65  38.880893 -76.929298  
16608  136609.78  38.897338 -77.002774  
16609  136784.11  38.898907 -76.985037  
16610  137321.65  38.903743 -77.042565  
16611  139826.42  38.926307 -77.042322

In [24]:
df[40:45] #sliced rows 40-45

X            Y       CCN              REPORT_DAT    SHIFT  \
40  407191.9434  136295.4350  25024345  2025/02/19 17:22:48+00      DAY   
41  398144.4800  138633.9700  25014552  2025/01/31 02:15:14+00  EVENING   
42  400130.5358  128889.0974  25014619  2025/01/31 02:59:13+00  EVENING   
43  400886.9908  133004.9529  25015121  2025/02/01 01:11:25+00  EVENING   
44  394088.4600  138708.7200  25015372  2025/02/01 16:06:25+00      DAY   

    METHOD              OFFENSE  \
40  OTHERS          THEFT/OTHER   
41  OTHERS          THEFT/OTHER   
42  OTHERS  MOTOR VEHICLE THEFT   
43  OTHERS  MOTOR VEHICLE THEFT   
44  OTHERS         THEFT F/AUTO   

                                              BLOCK         XBLOCK  \
40               5800 - 5899 BLOCK OF DIX STREET NE  407191.943365   
41                   620 - 699 BLOCK OF T STREET NW  398144.480000   
42               4200 - 4283 BLOCK OF 6TH STREET SE  400130.535806   
43  W STREET SE AND MARTIN LUTHER KING JR AVENUE SE  400886.990815   
44         1851 - 2008 BLOCK OF WISCONSIN AVENUE NW  394088.460000   

           YBLOCK   LATITUDE  LONGITUDE  
40  136295.435004  38.894477 -76.917096  
41  138633.970000  38.915570 -77.021395  
42  128889.097397  38.827787 -76.998497  
43  133004.952889  38.864864 -76.989780  
44  138708.720000  38.916226 -77.068165

In [25]:
df = df[(df['X']>=0) & (df['Y']>=0) & (df['CCN']>=0)] #filter data
df.describe()

X              Y           CCN         XBLOCK  \
count   16612.000000   16612.000000  1.661200e+04   16612.000000   
mean   399223.899504  137978.638332  2.509500e+07  399223.899504   
std      3011.298002    3482.734790  5.500669e+05    3011.298002   
min    390519.560000  127300.240000  5.131706e+06  390519.560000   
25%    397229.100000  136474.240300  2.503431e+07  397229.100000   
50%    398916.585000  137848.510000  2.507004e+07  398916.585006   
75%    401014.009600  139965.547500  2.510658e+07  401014.009567   
max    407806.750900  147250.570000  8.927285e+07  407806.750917   

              YBLOCK      LATITUDE     LONGITUDE  
count   16612.000000  16612.000000  16612.000000  
mean   137978.638332     38.909663    -77.008955  
std      3482.734790      0.031373      0.034721  
min    127300.240000     38.813473    -77.109338  
25%    136474.240252     38.896116    -77.031952  
50%    137848.510000     38.908480    -77.012499  
75%    139965.547500     38.927567    -76.988310  
max    147250.570000     38.993186    -76.910012

I selected CCN to be display, I sorted the data by CCN number, and filtered the data by ensuring the X, Y, and CCN columns were above 0